source/server/connection_handler_impl.cc

```diff
@@ -54,6 +66,30 @@ void ConnectionHandlerImpl::removeListeners(uint64_t listener_tag) {
   }
 }
 
+void ConnectionHandlerImpl::removeFilterChains(
+    const Network::DrainingFilterChains& draining_filter_chains, std::function<void()> completion) {
+  for (auto& listener : listeners_) {
+    // TODO(lambdai): merge the optimistic path and the pessimistic locking.
+    if (listener.second.listener_->listenerTag() ==
+        draining_filter_chains.getDrainingListenerTag()) {
+      ASSERT(listener.second.tcp_listener_.has_value());
+      listener.second.tcp_listener_->get().removeFilterChains(
+          draining_filter_chains.getDrainingFilterChains());
+      Event::DeferredTaskUtil::deferredRun(dispatcher_, std::move(completion));
+      return;
+    }
+  }
+  // Fallback to iterate over all listeners. The reason is that the target listener might have began
+  // another update and the previous tag is lost.
```

Comment:
This seems sub-optimal. Can you explain the flow here that causes this?

Reply:
Kind of, but this is pretty confusing as I need to page back in how all of the tagging stuff works, when a tag changes, etc. Can you add quite a bit more comments here to see if that helps me? Should the tag perhaps not update if we are doing a filter chain only update? I think that would make things a lot simpler?

Reply:
Yes but please add a TODO.

> Personally I feel the listener tag is helpful when I triage the listener update.

I think we can figure out a different way to help with debugging (SHA of proto, some other incrementing ID, etc.). By reusing the tag ID when a listener does not get reloaded I think it will make the logic easier to understand, but yes let's do that in a follow up. Thank you!